src/keystore/InMemoryKeystore.ts

```diff
@@ -294,10 +295,14 @@ export default class InMemoryKeystore implements Keystore {
           )
         }
 
+        const keyMaterial = getKeyMaterial(topicData.invitation)
+        const ciphertext = await encryptV2(payload, keyMaterial, headerBytes)
+
         return {
-          encrypted: await encryptV2(
-            payload,
-            getKeyMaterial(topicData.invitation),
+          encrypted: ciphertext,
+          senderHmac: await generateHmacSignature(
+            keyMaterial,
+            new TextEncoder().encode(contentTopic),
             headerBytes
```

Comment:
I think applying the hmac on the plaintext header bytes would allow for replays, if I'm understanding correctly. Would it make sense to use the ciphertext here?

Reply:
The header includes a timestamp which is verified against the envelope timestamp. That should limit the potential for replays. The header bytes are then used as AEAD for the message encryption, so you can't move the headers around without breaking decryption.

Reply:
The reason for using the header is that it's already serialized to bytes. The ciphertext contains has the payload, salt, and nonce. We'd have to pick just the payload, or maybe concatenate all of them. The headerBytes is a single thing to use in the HMAC which leaves less room for bugs or mistakes.